server/syntax.js

```diff
@@ -0,0 +1,100 @@
+'use strict';
+
+var _ = require('lodash');
+var j = require('junify');
+var esprima = require('esprima');
+
+Array.prototype.concatMap = function (f) {
+  return Array.prototype.concat.apply([], this.map(f));
+};
+
+function expressionsOf(ast, binding) {
+  return declarationsOf(ast).concatMap(function (node) {
+    if (j.unify({ type: 'FunctionDeclaration', id: identifier(binding), _: j._ }, node)) {
+      return [node.body];
+    } else if (j.unify({ type: 'VariableDeclarator', id: identifier(binding), _: j._ }, node)) {
+      return [node.init];
+    } else {
+      return [];
+    }
+  });
+}
+
+function expressionHasUsage(arg, target) {
+  return matches(arg, [
+    [{ type: 'CallExpression', callee: j.variable('sub'), _: j._ }, function (result) {
+      return expressionHasUsage(result.sub, target);
+    }],
+    [{ type: 'BinaryExpression', operator: j._, left: j.variable('sub1'), right: j.variable('sub2') }, function (result) {
+      return expressionHasUsage(result.sub1, target) || expressionHasUsage(result.sub2, target);
+    }],
+    [{ type: 'ReturnStatement', argument: j.variable('sub'), _: j._ }, function (result) {
+      return expressionHasUsage(result.sub, target);
+    }],
+    [{ type: 'ExpressionStatement', expression: j.variable('sub'), _: j._ }, function (result) {
+      return expressionHasUsage(result.sub, target);
+    }],
+    [{ type: 'BlockStatement', body: j.variable('sub'), _: j._ }, function (result) {
+      return result.sub.some(function (it) {
+        return expressionHasUsage(it, target);
+      });
+    }],
+    [{ type: 'VariableDeclaration', declarations: j.variable('sub'), _: j._ }, function (result) {
+      return result.sub.some(function (it) {
+        return expressionHasUsage(it, target);
+      });
+    }],
+    [{ type: 'VariableDeclarator', init: j.variable('sub'), _: j._ }, function (result) {
+      return expressionHasUsage(result.sub, target);
+    }],
+    [{ type: 'Identifier', name: j.variable('sub'), _: j._ }, function (result) {
+      return result.sub === target;
```

Comment:
yes, of course! I don't know why I commited this change. I was sure I had
it that way

On Mon, Jul 6, 2015 at 3:32 PM, Franco Leonardo Bulgarelli <
notifications@github.com> wrote:

> In server/syntax.js
> <https://github.com/mumuki/mumuki-mocha-server/pull/13#discussion_r33966661>
> :
>
> > +    }],
> > +    [{ type: 'BlockStatement', body: j.variable('sub'), _: j._ }, function (result) {
> > +      return result.sub.some(function (it) {
> > +        return expressionHasUsage(it, target);
> > +      });
> > +    }],
> > +    [{ type: 'VariableDeclaration', declarations: j.variable('sub'), _: j._ }, function (result) {
> > +      return result.sub.some(function (it) {
> > +        return expressionHasUsage(it, target);
> > +      });
> > +    }],
> > +    [{ type: 'VariableDeclarator', init: j.variable('sub'), _: j._ }, function (result) {
> > +      return expressionHasUsage(result.sub, target);
> > +    }],
> > +    [{ type: 'Identifier', name: j.variable('sub'), _: j._ }, function (result) {
> > +      return result.sub === target;
>
> Here we could just match the expected name and return _.constant(true),
> right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mumuki/mumuki-mocha-server/pull/13/files#r33966661>.
>